### ISSUE TITLE
How to shuffle favorites

### ISSUE THREAD

**Issue by m3lvn** [jishi/node-sonos-http-api#434]
Hi,

First of all thanks for this great API. I am relatively new to the Raspberry but I have managed to have my Nest, TV and Sonos working with Homebridge. However, is there a guide available to shuffle a favorite playlist? I have a switch in my iOS app that sends the HTTP request: 

`localhost:5000/name of room/favorite/name of favorite playlist`

Everything works just fine, but how do I make sure that the playlist is played in shuffle mode? I don't want to start every single day with the same song when I tell Siri "Goodmorning"

I don't get the instructions on what files to change in all the different locations.

Thanks!

**Comment by jishi** [comment]
Since selecting a favorite actually autostarts, you must set shuffle to on before you select a favorite for playback, otherwise it will always start with the same track.

However, you can only set shuffle, if the player is playing from the queue (which wouldn't be the case if you are playing a radio station).

What you could test, is to define a preset, because then you can specify both play modes as well as a favorite, something like (create the file morning.json in the presets folder):

`/presets/morning.json`

```
{
  "players": [{ "roomName": "name of room"}],
  "favorite": "name of favorite playlist",
  "playMode": { "shuffle": true }
}
```

It might be that presets also starts before it has applied playMode. Another option would be to create your own custom favorite action, that doesn't invoke play. Then you could load favorite, enable shuffle, and then invoke play manually. 

Basically, just copy https://github.com/jishi/node-sonos-http-api/blob/master/lib/actions/favorite.js and put the copy in the same folder, and remove line 4. Adjust line 8 and 9 and define the new action name (and maybe remove one of them since you don't care about different spellings).

**Comment by rafale77** [comment]
Don't mean to hijack the issue but is there a way (either through preset or a command) to have the sonos go back to the playing the queue instead either line-in or a radio station?

**Comment by jishi** [comment]
@rafale77 Your question is unclear (do you want to go back to the queue, or line-in? or radio station?), and I suggest you open a new issue instead, since any reply you make here emails both the original author of the issue and me. It gets kind of annoying for people :)

**Comment by rafale77** [comment]
I want to go back to the queue while the player is currently playing either a station or line-in.

**Comment by m3lvn** [comment]
Thanks for your swift response. I just tested `/presets/morning` but then the browser only states:
`["example","morning"]`

I have no idea what you mean with the second part. 

> and remove line 4. Adjust line 8 and 9 and define the new action name

Where do I start? I found the location of the favorite.js file but what do I need to add to line 8 and 9?
Situation: I have a favorite in Sonos called: "Easy on Sunday". I usually go to: `localhost:5000/woonkamer/favorite/easy%20on%20sunday/` and then the it starts paying.

Now I want that playlist to start playing in shuffle-mode.

Thanks for your help!

**Comment by jishi** [comment]
not "presets", that gives you a listing of all presets. It should be singular "preset".

If you copied favorites.js, change this section:

```
module.exports = function (api) {
  api.registerAction('favorite', favorite);
  api.registerAction('favourite', favorite);
}`
```

to 
```
module.exports = function (api) {
  api.registerAction('favorite-without-play', favorite);
}
```

restart the api, and then you should be able to call `/PlayerName/favorite-without-play/easy%20on%20sunday` which would queue the favorite, `/PlayerName/shuffle/on` to turn on shuffle, then `/PlayerName/play` to start playback.

But if the preset route actually works, that would be much simpler.

**Comment by m3lvn** [comment]
This is great! The singular "preset" did the job. I have added multiple presets which play in shuffle mode.

Just one more question. Whilst playing a Sonos playlist, is there a way to switch to TV-source from the Playbar? Or is that off-topic?

Thanks a million for your help!

**Comment by jishi** [comment]
Yes there is. It's not as convenient as the line-in action (I plan on adding support for PLAYBAR, eventually), but it's possible, see https://github.com/jishi/node-sonos-http-api/issues/382

**Comment by m3lvn** [comment]
Thanks! What about the volume in presets? Whats the code and where to add it?

**Comment by jishi** [comment]
You add volume to the same object that specifies roomName: 

```
{
  "players": [{ "roomName": "name of room", "volume": 30}],
  "favorite": "name of favorite playlist",
  "playMode": { "shuffle": true }
}
```

You can see examples in the README.
